lib/stm32/f4/rcc.c

```diff
@@ -569,22 +621,48 @@ void rcc_set_rtcpre(uint32_t rtcpre)
 }
 
 void rcc_set_main_pll_hsi(uint32_t pllm, uint32_t plln, uint32_t pllp,
-			  uint32_t pllq)
+			  uint32_t pllq, uint32_t pllr)
 {
-	RCC_PLLCFGR = (pllm << RCC_PLLCFGR_PLLM_SHIFT) |
+	uint32_t reg32 = RCC_PLLCFGR;
+
+	/* mask out any previous values */
+	reg32 &= ~(
+		 RCC_PLLCFGR_PLLSRC |	/* 0 = HSI */
+		(RCC_PLLCFGR_PLLM_MASK << RCC_PLLCFGR_PLLM_SHIFT) |
+		(RCC_PLLCFGR_PLLN_MASK << RCC_PLLCFGR_PLLN_SHIFT) |
+		(RCC_PLLCFGR_PLLP_MASK << RCC_PLLCFGR_PLLP_SHIFT) |
+		(RCC_PLLCFGR_PLLQ_MASK << RCC_PLLCFGR_PLLQ_SHIFT) |
+		(RCC_PLLCFGR_PLLR_MASK << RCC_PLLCFGR_PLLR_SHIFT));
```

Comment:
What about SOIC without an R. Please look at the PR I did - it has 2 reserved values and if R is passed as 0 (a non legal value) it will not change the default value - this code will, and if passed 0 it will use the illegal value.